# Đồ án cuối kỳ môn Lập trình cho Khoa học dữ liệu
## Giảng viên hướng dẫn: 
### Thầy Trần Trung Kiên
### Cô Phan Thị Phương Uyên
## Sinh viên thực hiện:
##### Dương Hoàng Quang Khanh, mã sinh viên: 1712521

---

## Import các thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## Về nguồn dữ liệu:
Dữ liệu được sử dụng trong bài tập này là dữ liệu về 20 nhóm mặt hàng xuất khẩu và 20 nhóm mặt hàng xuất khẩu chủ lực của Việt Nam với các quốc gia trên thế giới trong giai đoạn 2010-2019. Dữ liệu được thống kê bởi Phòng ban thống kê Liên Hợp Quốc (United Nations Statistics Division) và được em trích xuất và tải về từ trang https://comtrade.un.org/data. Dữ liệu được thống kê theo từng năm và được lưu trong file "comtrade_data.csv". 

## Khám phá dữ liệu
### Đọc dữ liệu từ file

In [2]:
pd.set_option('display.max_columns', None) #Giúp hiển thị tất cả các cột của dữ liệu
data = pd.read_csv("comtrade_data.csv")
data.head(10)

Classification  Year  Period  Period Desc.  Aggregate Level  \
0             H5  2019    2019          2019                2   
1             H5  2019    2019          2019                2   
2             H5  2018    2018          2018                2   
3             H4  2017    2017          2017                2   
4             H4  2016    2016          2016                2   
5             H5  2019    2019          2019                2   
6             H4  2016    2016          2016                2   
7             H4  2015    2015          2015                2   
8             H5  2018    2018          2018                2   
9             H3  2011    2011          2011                2   

   Trade Flow Code Trade Flow  Reporter Code  Reporter Reporter ISO  \
0                0          M            704  Viet Nam          VNM   
1                0          M            704  Viet Nam          VNM   
2                0          M            704  Viet Nam          VNM   
3                0          M            704  Viet Nam          VNM   
4                0          M            704  Viet Nam          VNM   
5                0          M            704  Viet Nam          VNM   
6                0          M            704  Viet Nam          VNM   
7                0          M            704  Viet Nam          VNM   
8                0          M            704  Viet Nam          VNM   
9                0          M            704  Viet Nam          VNM   

   Partner Code    Partner Partner ISO  Commodity Code  \
0           842        USA         USA               2   
1           699      India         IND               2   
2           842        USA         USA               2   
3           842        USA         USA               2   
4           699      India         IND               2   
5            36  Australia         AUS               2   
6           842        USA         USA               2   
7           842        USA         USA               2   
8           699      India         IND               2   
9           842        USA         USA               2   

                    Commodity  Qty Unit Code  Qty Unit  Qty  \
0  Meat and edible meat offal             -1       NaN  0.0   
1  Meat and edible meat offal             -1       NaN  0.0   
2  Meat and edible meat offal             -1       NaN  0.0   
3  Meat and edible meat offal             -1       NaN  0.0   
4  Meat and edible meat offal             -1       NaN  0.0   
5  Meat and edible meat offal             -1       NaN  0.0   
6  Meat and edible meat offal             -1       NaN  0.0   
7  Meat and edible meat offal             -1       NaN  NaN   
8  Meat and edible meat offal             -1       NaN  0.0   
9  Meat and edible meat offal             -1       NaN  NaN   

   Alt Qty Unit Code  Alt Qty Unit  Alt Qty  Netweight (kg)  \
0                 -1           NaN      0.0             0.0   
1                 -1           NaN      0.0             0.0   
2                 -1           NaN      0.0             0.0   
3                 -1           NaN      0.0             NaN   
4                 -1           NaN      0.0             NaN   
5                 -1           NaN      0.0             0.0   
6                 -1           NaN      0.0             NaN   
7                 -1           NaN      NaN             NaN   
8                 -1           NaN      0.0             0.0   
9                 -1           NaN      NaN             NaN   

   Gross weight (kg)  Trade Value (US$)  CIF Trade Value (US$)  \
0                0.0          201058755           2.010588e+08   
1                0.0          178868209           1.788682e+08   
2                0.0          168236990           1.682370e+08   
3                NaN          123859577           1.238596e+08   
4                NaN          103212633           1.032126e+08   
5                0.0           98907346           9.890735e+07   
6              

### Kích thước của dữ liệu

In [3]:
print(data.shape)

(31520, 27)


Có thể thấy dữ liệu gồm có 31520 dòng và 28 cột.

Mỗi dòng dữ liệu mô tả về giá trị xuất khẩu hoặc nhập khẩu của Việt Nam với một đối tác về một nhóm mặt hàng trong một năm duy nhất.

Về mô tả thông tin của mỗi cột xin mời xem mô tả dưới đây:

- **Classification**: các phân loại quốc tế áp dụng cho dữ liệu thương mại và thuế quan được đưa ra bởi tổ chức Hải quan thế giới (WCO), H3 áp dụng cho hàng hóa từ 2007 - 2012, H4 là từ 2012 - 2017, và H5 là từ 2017 đến nay.
- **Year, Period, Period Desc**: Vì dữ liệu thông kê theo từng năm nên cả ba cột này đều mô tả năm mà dữu liệu được thống kê.
- **Aggregate Level**: đề cập đến mức độ chi tiết trong thống kê dữ liệu.
- **Trade Flow Code, Trade Flow**: phân biệt giữa mặt hàng xuất khẩu và nhập khẩu. 
- **Reporter Code, Reporter, Reporter ISO**: lần lượt là Mã số, Tên, Viết tắt của quốc gia được chọn là chủ thể thu thập thông tin. Ở đây là Việt Nam.
- **Partner Code, Partner, Partner ISO**: lần lượt là Mã số, Tên Viết tắt của các quốc gia đối tác.
- **Qty Unit Code, Qty Unit, Qty, Alt Qty Unit Code, Alt Qty Unit, Alt Qty, Netweight (kg), Gross weight (kg)**: Các cột này nói về các đơn vị số lượng và khối lượng hàng hóa, nhưng vì ở đây người ta chỉ thống kê về giá trị nên cá cột này không mang ý nghĩa.
- **Trade Value, CIF Trade Value, FOB Trade Value**: được tính theo US$, là tổng giá trị trong năm đó.

## Tiền xử lý dữ liệu

### Loại bỏ trước một số cột không mang lại giá trị thông tin

Các giá trị trong một số cột mà ta xem xét sẽ bỏ.

In [4]:
data["Classification"].unique()
c = ["Classification", "Reporter", "Reporter Code", "Reporter ISO", "Qty Unit Code",
     "Qty Unit", "Qty", "Alt Qty Unit Code", "Alt Qty Unit", "Alt Qty", "Netweight (kg)", "Gross weight (kg)"]
unq = pd.Series([data[i].unique() for i in c], index=c)
unq

Classification       [H5, H4, H3]
Reporter               [Viet Nam]
Reporter Code               [704]
Reporter ISO                [VNM]
Qty Unit Code                [-1]
Qty Unit                    [nan]
Qty                    [0.0, nan]
Alt Qty Unit Code            [-1]
Alt Qty Unit                [nan]
Alt Qty                [0.0, nan]
Netweight (kg)         [0.0, nan]
Gross weight (kg)      [0.0, nan]
dtype: object

- Cột **"Classification"** này chỉ có 3 giá trị là 'H3', 'H4', 'H5'. Nó biểu thị cho hàng hóa giao dịch trong giai đoạn nào ứng với 3 giá trị này. Vì trong dữ liệu này ta đã có thông tin về năm giao dịch ở cột **"Year"**, thậm chí cột này còn chi tiết về năm hơn, do đó ta có thể bỏ đi cột **"Classification"**.
- Ba cột **"Reporter", "Reporter ISO", "Partner Code"** đều chỉ chứa 1 giá trị trong mỗi cột (liên quan đến Việt Nam) vì trong dữ liệu này thì Việt Nam là quốc gia được thống kê, do đó ta có thể bỏ giữ lại 1 trong 3 cột này.
- Các cột **Qty Unit Code, Qty Unit, Qty, Alt Qty Unit Code, Alt Qty Unit, Alt Qty, Netweight (kg), Gross weight (kg)** chỉ chứ các giá trị không có ý nghĩa nên ta cũng sẽ bỏ.
- Ba cột là **"Year", "Period", "Period Desc."** đều cùng mang một giá trị là năm thống kê nên ta chỉ cần giữ lại một cột là **"Year"** và bỏ đi hai cột **"Period", "Period Desc."**.
- Hai cột cùng biểu diễn các giá trị phân biệt Xuất khẩu với nhập khẩu là **"Trade Flow Code"**và **"Trade Flow"** chỉ cần giữ lại 1 cột.
- Các cột **"Partner Code", "Partner", "Partner ISO"** đều biểu diễn cùng một quốc gia đối tác trên một dòng, nên ta có thể bỏ đi hai cột là **"Partner Code", "Partner ISO"** và giữ lại cột **"Partner""**
- Một điều đáng lưu ý Mới các mặt hàng nhập khẩu thì **"Trade value"** sẽ bằng **"CIF Trade Value"**, ngược lại với hàng hóa xuất khẩu thì **"Trade Value"** sẽ bằng **"FOB Trade Value"**. Dó đó ta chỉ cẩn giữ lại cột **Trade Value**.

Các cột ta sẽ bỏ đi đó là "Classification", "Reporter", "Reporter Code", "Reporter ISO", "Qty Unit Code", "Qty Unit", "Qty", "Alt Qty Unit Code", "Alt Qty Unit", "Alt Qty", "Netweight (kg)", "Gross weight (kg)","Period", "Period Desc.", "Partner Code", "Partner ISO", "CIF Trade Value", "FOB Trade Value".

In [5]:
drop_cols = ["Classification", "Trade Flow Code", "Reporter Code", "Reporter ISO", "Qty Unit Code",
             "Qty Unit", "Qty", "Alt Qty Unit Code", "Alt Qty Unit", "Alt Qty", "Netweight (kg)", "Gross weight (kg)", 
             "Period", "Period Desc.", "Partner Code", "Partner ISO", "CIF Trade Value (US$)", "FOB Trade Value (US$)"]
data.drop(labels=drop_cols, axis=1, inplace=True)
data.head(10)

Year  Aggregate Level Trade Flow  Reporter    Partner  Commodity Code  \
0  2019                2          M  Viet Nam        USA               2   
1  2019                2          M  Viet Nam      India               2   
2  2018                2          M  Viet Nam        USA               2   
3  2017                2          M  Viet Nam        USA               2   
4  2016                2          M  Viet Nam      India               2   
5  2019                2          M  Viet Nam  Australia               2   
6  2016                2          M  Viet Nam        USA               2   
7  2015                2          M  Viet Nam        USA               2   
8  2018                2          M  Viet Nam      India               2   
9  2011                2          M  Viet Nam        USA               2   

                    Commodity  Trade Value (US$)  Flag  
0  Meat and edible meat offal          201058755     4  
1  Meat and edible meat offal          178868209     4  
2  Meat and edible meat offal          168236990     4  
3  Meat and edible meat offal          123859577     0  
4  Meat and edible meat offal          103212633     0  
5  Meat and edible meat offal           98907346     4  
6  Meat and edible meat offal           97801353     0  
7  Meat and edible meat offal           95843013     0  
8  Meat and edible meat offal           86784142     4  
9  Meat and edible meat offal           86117512     0

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31520 entries, 0 to 31519
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Year               31520 non-null  int64 
 1   Aggregate Level    31520 non-null  int64 
 2   Trade Flow         31520 non-null  object
 3   Reporter           31520 non-null  object
 4   Partner            31520 non-null  object
 5   Commodity Code     31520 non-null  int64 
 6   Commodity          31520 non-null  object
 7   Trade Value (US$)  31520 non-null  int64 
 8   Flag               31520 non-null  int64 
dtypes: int64(5), object(4)
memory usage: 2.2+ MB


Nhìn chung do quá trình thống kê của United Nations Statistics Division là rất kỹ lưỡng nên ngoài không có giá trị nào bị thiếu.

## Quay lại bước khám phá dữ liệu

### Mỗi cột có kiểu dữ liệu gì?

In [7]:
col_dtypes = data.dtypes
col_dtypes

Year                  int64
Aggregate Level       int64
Trade Flow           object
Reporter             object
Partner              object
Commodity Code        int64
Commodity            object
Trade Value (US$)     int64
Flag                  int64
dtype: object

Ta thấy ngoài các cột có dữ liệu kiểu int thì có 4 cột có dữ liệu là kiểu object. Để xem các cột này có phải chỉ chứa kiểu dữ liệu chuỗi không ta thực hiện như dưới đây:

In [8]:
object_cols = ["Trade Flow", "Reporter","Partner", "Commodity"]
for c in object_cols:
    dtypes = set()
    for i in range(len(data[c])):
        dtypes.add(type(data.loc[i,c]))
    print("Column ",c," has type: ", dtypes)

Column  Trade Flow  has type:  {<class 'str'>}
Column  Reporter  has type:  {<class 'str'>}
Column  Partner  has type:  {<class 'str'>}
Column  Commodity  has type:  {<class 'str'>}


Như vậy là các cột này đều chỉ có kiểu string.

### Các cột dữ liệu dạng categorical có những giá trị nào?

In [9]:
NumUniqueVals = pd.Series([len(data[i].unique()) for i in ["Trade Flow", "Reporter","Partner", "Commodity", "Commodity Code"]],
                          index= ["Trade Flow", "Reporter","Partner", "Commodity", "Commodity Code"])
NumUniqueVals
# data[["Partner"]].value_counts()

Trade Flow          2
Reporter            1
Partner           176
Commodity          34
Commodity Code     34
dtype: int64

In [10]:
print(data["Trade Flow"].unique())

['M' 'X']


Dữ liệu ở cột **Trade Flow** chỉ có hai giá trị là 'X' và 'M' tương ứng với xuất khẩu (Export) và nhập khẩu (Import). Hai giá trị này có thể sẽ được dùng để tách dữ liệu làm 2 DataFrame cho hàng hóa xuất - nhập khẩu riêng biệt.


## Đưa ra các câu hỏi cần trả lời từ dữ liệu:

- Câu 1: Có 20 nhóm hàng xuất khẩu và 20 nhóm hàng nhập khẩu nhưng tại sao chỉ có 34 gián trị khác nhau ở cột **Commodity?** Như vậy phải có 6 loại hàng hóa nằm ở cả hai danh mục xuất nhập khẩu. Đó là những loại hàng nào, có mã là gì? 6 loại hàng hóa này có mang ý nghĩa gì khi nằm ở cả hai hạng mục.
- Câu 2: Trong giai đoạn thống kê trên thì đâu là những quốc gia là đối tác lớn nhất của Việt Nam ở cả thị trường xuất khẩu và nhập khẩu? Nêu ra 3 quốc gia.
- Câu 3: Tính theo từng năm hãy chỉ ra 3 nhóm mặt hàng xuất khẩu chủ lực và 3 nhóm mặt hàng nhập khẩu chính.
- Câu 4: Trong những năm này thì năm nào Việt Nam là quốc gia Xuất siêu hay Nhập siêu.

### Tách dữ liệu:
Để tiện cho việc trả lời các câu hỏi được nêu thì đầu tiên ta tách dữ liệu thành hai DataFrame riêng biệt, một DataFrame lưu các thông tin về hàng hóa xuất khẩu, DataFrame kia lưu thông tin về hàng hóa nhập khẩu.

In [48]:
export_df = data[data["Trade Flow"] == 'X']
export_df=export_df.drop(labels="Trade Flow", axis=1)
export_df.head()


Year  Aggregate Level  Reporter Partner  Commodity Code  \
345  2019                2  Viet Nam   China               3   
346  2014                2  Viet Nam     USA               3   
347  2017                2  Viet Nam   China               3   
348  2018                2  Viet Nam     USA               3   
349  2018                2  Viet Nam   China               3   

                                             Commodity  Trade Value (US$)  \
345  Fish and crustaceans, molluscs and other aquat...         1214590404   
346  Fish and crustaceans, molluscs and other aquat...         1063638814   
347  Fish and crustaceans, molluscs and other aquat...         1061198627   
348  Fish and crustaceans, molluscs and other aquat...         1048403135   
349  Fish and crustaceans, molluscs and other aquat...          998828981   

     Flag  
345     4  
346     0  
347     0  
348     4  
349     4

In [47]:
import_df = data[data["Trade Flow"] == 'M']
import_df=import_df.drop(labels="Trade Flow", axis=1)
import_df.head()

Year  Aggregate Level  Reporter Partner  Commodity Code  \
0  2019                2  Viet Nam     USA               2   
1  2019                2  Viet Nam   India               2   
2  2018                2  Viet Nam     USA               2   
3  2017                2  Viet Nam     USA               2   
4  2016                2  Viet Nam   India               2   

                    Commodity  Trade Value (US$)  Flag  
0  Meat and edible meat offal          201058755     4  
1  Meat and edible meat offal          178868209     4  
2  Meat and edible meat offal          168236990     4  
3  Meat and edible meat offal          123859577     0  
4  Meat and edible meat offal          103212633     0

### Câu 1:
Những nhóm hàng nào vừa nằm ở danh mục sản phẩm xuất khẩu và nhập khẩu, chúng có mã là gì? 6 loại hàng hóa này có mang ý nghĩa gì khi nằm ở cả hai hạng mục.

In [49]:
ex_commodity_set = set(export_df["Commodity"].unique())
im_commodity_set = set(import_df["Commodity"].unique())
intersect_commodity = ex_commodity_set.intersection(im_commodity_set)
for i in intersect_commodity:
    print("- ", i)
ex_commoditycode_set = set(export_df["Commodity Code"].unique())
im_commoditycode_set = set(import_df["Commodity Code"].unique())
intersect_commoditycode = ex_commoditycode_set.intersection(im_commoditycode_set)
print("Mã các loại mặt hàng: ",intersect_commoditycode)

-  Glass and glassware
-  Electrical machinery and equipment and parts thereof; sound recorders and reproducers; television image and sound recorders and reproducers, parts and accessories of such articles
-  Plastics and articles thereof
-  Mineral fuels, mineral oils and products of their distillation; bituminous substances; mineral waxes
-  Furniture; bedding, mattresses, mattress supports, cushions and similar stuffed furnishings; lamps and lighting fittings, n.e.c.; illuminated signs, illuminated name-plates and the like; prefabricated buildings
-  Footwear; gaiters and the like; parts of such articles
Mã các loại mặt hàng:  {64, 70, 39, 85, 27, 94}


- Như vậy 6 loại mặt hàng được vừa có mặt trong danh mục nhập khẩu và xuất khẩu là 6 nhóm mặt hàng được kể ra ở trên, ứng với các "Commodity Code" là: 64, 70, 39, 85, 27, 94.
- Việc có một số nhóm sản phàm nằm ở cả hai danh mục trên theo em có thể được giải thích một cách dễ hiểu như sau:
    + Thứ nhất, đây là nhóm mặt hàng gồm có nhiều dòng sản phẩm, chúng ta xuất khẩu một số dòng sản phẩm và nhập khẩu về các loại sản phẩm khác trong nhóm này. Ví dụ rõ nhất là hai nhóm hàng liên quan đến các sản phẩm nhựa (mã 39) và dày dép (mã 64) ở trên, ở Việt Nam có nhiều công ty lớn sẳn xuất hai mặt hàng này và chúng được xuất khuẩ sang các nước, đồng thời chúng ta vần nhập khẩu các sản phẩm nhựa hay giày dép từ các nước bên ngoài.
    + Thứ hai, đây là các nhóm mặt hàng được tái xuất khẩu hoặc tái nhập khẩu. Điển hình là các mặt hàng thuộc nhóm Chất khoáng và dầu mỏ (mã 27). Chúng ta xuất khẩu dầu thô sáng các nước vừa nhập dầu thô và xăng thành phẩm về. Nguyên nhân là do:  Kết quả từ phòng thí nghiệm cho thấy sản phẩm dầu thô của Bạch Hổ có hàm lượng lưu huỳnh rất thấp, được thị trường nước ngoài ưa chuộng, có thể bán với giá cao. Đồng thời cũng kết quả thí nghiệm cho thấy chúng ta có thể pha 20% dầu thô nhập từ nước ngoài với giá thấp mà vẫn cho ra sản phẩm tốt. Vì thế mới có chuyện xuất dầu thô mà vẫn nhập dầu thô.

### Câu 2: 
Trong giai đoạn thống kê trên thì đâu là những quốc gia là đối tác lớn nhất của Việt Nam ở cả thị trường xuất khẩu và nhập khẩu? Nêu ra 3 quốc gia.

In [55]:
export_df[export_df["Commodity Code"]==64]

Year  Aggregate Level  Reporter                  Partner  \
19395  2019                2  Viet Nam                      USA   
19396  2018                2  Viet Nam                      USA   
19397  2017                2  Viet Nam                      USA   
19398  2016                2  Viet Nam                      USA   
19399  2015                2  Viet Nam                      USA   
...     ...              ...       ...                      ...   
20983  2016                2  Viet Nam   Dem. Rep. of the Congo   
20984  2016                2  Viet Nam  United Rep. of Tanzania   
20987  2017                2  Viet Nam                    Benin   
20992  2015                2  Viet Nam                  Iceland   
20993  2015                2  Viet Nam                    Kenya   

       Commodity Code                                          Commodity  \
19395              64  Footwear; gaiters and the like; parts of such ...   
19396              64  Footwear; gaiters and the like; parts of such ...   
19397              64  Footwear; gaiters and the like; parts of such ...   
19398              64  Footwear; gaiters and the like; parts of such ...   
19399              64  Footwear; gaiters and the like; parts of such ...   
...               ...                                                ...   
20983              64  Footwear; gaiters and the like; parts of such ...   
20984              64  Footwear; gaiters and the like; parts of such ...   
20987              64  Footwear; gaiters and the like; parts of such ...   
20992              64  Footwear; gaiters and the like; parts of such ...   
20993              64  Footwear; gaiters and the like; parts of such ...   

       Trade Value (US$)  Flag  
19395         6676322579     4  
19396         5834537664     4  
19397         5126498739     0  
19398         4506471137     0  
19399         4097425481     0  
...                  ...   ...  
20983                300     0  
20984                300     0  
20987                250     0  
20992                216     0  
20993                210     0  

[1105 rows x 8 columns]